v7/_posts/odata features/2017-12-26-06-00-JsonBatch.md

```diff
@@ -0,0 +1,330 @@
+                    }
+                    else if (operationMessage.Method == "PATCH")
+                    {
+                        // Process flow of PATCH method similar to PUT method above.
+                        // ...
+                    }
+                    // additional else if blocks for other HTTP method processs flow come here...
+
+                break;
+
+                case ODataBatchReaderState.ChangesetStart:
+
+                    // Request's GroupId can be passed to response writer for correlation of groupId between request & response.												
+                    batchWriter.WriteStartChangeset(batchReader.CurrentGroupId);
+
+                    // Handle atomic group starting here as needed by implementation of specific services.
+                    ...
+                    // End of atomic group start processing.
+                    break;
+
+                case ODataBatchReaderState.ChangesetEnd:
+
+                    batchWriter.WriteEndChangeset();
+
+                    // Handle atomic group ending here as needed by implementation of specific services.
+                    ...
+                    // End of atomic group end processing.
+                    break;
+            }
+        }
+        
+        batchWriter.WriteEndBatch();
+{% endhighlight %}
+
+With the introduction of JSON batch, the following public APIs / attributes are available:
+* `ODataBatchWriter`
+    * `public ODataBatchOperationRequestMessage CreateOperationRequestMessage(string method, System.Uri uri, string contentId, BatchPayloadUriOption payloadUriOption, IEnumerable<string> dependsOnIds)`: This can be used to create operation request message with the additional dependsOnIds that explicitly specify operation dependencies in a JSON batch request. For Multipart/Mixed batch, this new overloaded API can also be used, as well as the two original APIs without the need to explicitly specify the dependsOnIds, which will be derived implicitly per the protocol.
+    * `public void WriteStartChangeset(string)`: This can be used to specify the atomic group Id when the operation request is created, or the operation response's atomic group Id correlating to value in the request as required by the OData JSON Batch semantics.
+* `ODataBatchReader`:
+	* `public string CurrentGroupId`: This returns the current group Id when the atomic group is created. It can be used to create the same group Id on the batch response corresponding to that in the incoming request, as required by  the OData JSON Batch semantics. Please note that this is not the case for Multipart/Mixed format where the changeset boundary string in request and response are not correlated.
+* `ODataBatchOperationRequestMessage`:
+	* `public string GroupId`: This returns the operation request's group Id. For JSON batch, this is the value specified by the user; For Multipart/Mixed, this is the GUID used to generated the boundary parameter value of the changeset.
```

Comment:
Is this always a generated GUID? I thought that that the user could explicitly specify the boundary, and we just fell back to GUID generation if the user didn't define the boundary.

Reply:
yes you're correct. Not necessary a GUID. Updated to "Id string".